src/rdkafka.h

```diff
@@ -6225,6 +6231,350 @@ rd_kafka_oauthbearer_set_token_failure (rd_kafka_t *rk, const char *errstr);
 
 /**@}*/
 
+
+/**
+ * @name Transactional producer API
+ *
+ * The transactional producer operates on top of the idempotent producer,
+ * and provides full exactly-once semantics (EOS) for Kafka when used with
+ * the transaction aware consumer (\c isolation.level=read_committed).
+ *
+ * A producer instance is configured for transactions by setting the
+ * \c transactional.id to an identifier unique for the application. This
+ * id will be used to fence stale transactions from previous instances of
+ * the application, typically following an outage or crash.
+ *
+ * After creating the transactional produce instance using rd_kafka_new()
+ * the transactional state must be initialized by calling
+ * rd_kafka_init_transactions(). This is a blocking call that will
```

Comment:
we should perhaps start accommodating cancellable versions of sync calls (no time like the present?)

To do this, you'd need variants of the sync methods that return immediately and signal their completion via an event on a queue. you then need some way for the application to signal to librdkafka a cancellation is desired and a mechanism for librdkafka to pick up on that. there are various ways that could be set up. for example, you could pass in a cancellation token similar to .net (see `CancellationTokenSource` + `CancellationToken` classes). Or `rd_kafka_init_transactions` could be responsible for generating a token, returning it, and then if the application wants to cancel the operation, it uses the returned token to pass into a `rd_kafka_init_transactions_cancel` call. there's precedent for this approach too - e.g. timers in javascirpt work somewhat like this.

my instincts at the moment are erring on something like the second approach. you'd want to avoid allocs, so the token could perhaps be an incrementing id.

Reply:
I think the queue interface makes most sense since it is what we use in other places, and it is very flexible and caters to all kinds of asynchronous needs.
I.e., you can use it to wait on a single object (the queue) for multiple outstanding operations, you can cancel it at any time by destroying the queue, you can forward the results to another queue, etc.

But, all this is outside the scope of the initial API.

Reply:
sounds good.